examples/pytorch/gcn.py

```diff
@@ -0,0 +1,210 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""Example of GCN model trained on Cora dataset."""
+
+from typing import Tuple
+from dataclasses import dataclass
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch_geometric.nn import GCN2Conv
+from torch_geometric.nn.conv.gcn_conv import gcn_norm
+from torch_sparse.tensor import SparseTensor
+
+from deepgnn.graph_engine import Graph, graph_ops, get_logger
+from deepgnn.graph_engine.data.citation import Cora
+
+
+@dataclass
+class GCNQueryParameter:
+    """Configuration for graph query to fetch data for model."""
+
+    neighbor_edge_types: np.ndarray
+    feature_idx: int
+    feature_dim: int
+    label_idx: int
+    label_dim: int
+    num_classes: int
+    num_hops: int = 2
+
+
+def query(
+    inputs: np.ndarray, graph: Graph, param: GCNQueryParameter
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Fetch graph data."""
+    nodes, edges, _ = graph_ops.sub_graph(
+        graph,
+        inputs,
+        edge_types=param.neighbor_edge_types,
+        num_hops=param.num_hops,
+        self_loop=True,
+        undirected=True,
+        return_edges=True,
+    )
+
+    # Extract features with a single call to graph engine.
+    feat_labels = graph.node_features(
+        nodes,
+        np.array(
+            [
+                [param.feature_idx, param.feature_dim],
+                [param.label_idx, param.label_dim],
+            ],
+            dtype=np.int32,
+        ),
+        np.float32,
+    )
+
+    feat_labels = torch.from_numpy(feat_labels)
+    adj = SparseTensor.from_edge_index(
+        torch.from_numpy(np.transpose(edges)),
+    )
+
+    return (
+        feat_labels[:, : param.feature_dim],
+        gcn_norm(adj),
+        feat_labels[:, param.feature_dim : param.feature_dim + param.label_dim]
+        .to(torch.int64)
+        .squeeze(),
+    )
+
+
+class GCN(nn.Module):
+    """
+    Example GCN model from PyG.
+
+    Original example:
+    https://github.com/pyg-team/pytorch_geometric/blob/02cc18d9d6841ecda4eb61eebb48b86f1aaae477/examples/gcn2_ppi.py#L23
+    """
+
+    def __init__(
+        self,
+        hidden_channels: int,
+        num_layers: int,
+        alpha: float,
+        theta: float,
+        feature_dim: int,
+        num_classes: int,
+        shared_weights: bool = True,
+        dropout: float = 0.0,
+    ):
+        """Initialize model."""
+        super().__init__()
+
+        self.lins = nn.ModuleList()
+        self.lins.append(nn.Linear(feature_dim, hidden_channels))
+        self.lins.append(nn.Linear(hidden_channels, num_classes))
```

Comment:
Can pull these appends into the constructor of the modulelist.

Reply:
Sure, it makes everything longer.